### PR TITLE
Google Ads: Offline conversions import destination

### PR DESCRIPTION
**Pull request by gmcro**
Merge the offline conversions import code.

1 potential issue: We may need to make sure the Google Ads package gets installed properly an set a fixed version number.

### CODE REVIEW

File: airflow_requirements.txt [hunk 1]
```diff
@@ -8,4 +8,5 @@ pydantic==1.10.9
 # Data sources requirements
 apache-airflow-providers-apache-drill==2.4.3
 google-cloud-bigquery==3.11.3
+google-ads
```
Comment (caiotomazelli) [inline airflow_requirements.txt:11]: Are you explicitly choosing not to add a specific version for this? Consider adding at least a compatible release requirement, e.g. google-ads==22.* or google-ads>=22.0

File: dags/destinations/gads_oci.py [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+ConversionIndicesToConversions = List[Tuple[int, Any]]
+CustomerConversionMap = Dict[str, ConversionIndicesToConversions]
+InvalidConversionIndices = List[Tuple[int, errors.ErrorNameIDMap]]
+PartialFailures = Dict[int, str]
+
+
+class Destination:
+  """Implements DestinationProto protocol for Google Ads OCI."""
+
+  def __init__(self, config: Dict[str, Any]) -> None:
```
Comment (caiotomazelli) [inline dags/destinations/gads_oci.py:51]: Remove "-> None" to match the [DestinationProto](https://github.com/google-marketing-solutions/Tightlock/blob/main/dags/protocols/destination_proto.py) signature. I think this is the cause for the failing test.

File: dags/destinations/gads_oci.py [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+    For example:
+      valid_conversions = {"customer1": [(1, "ConversionOne"), (2, "Conversion2")]}
+      invalid_indices_and_errors = [3, MISSING_MADATORY_FIELDS, ...]
+    """
+    valid_conversions = defaultdict(list)
+    invalid_indices_and_errors = []
+
+    for i, conversion in enumerate(offline_conversions):
+      valid = True
+      print(conversion)
```
Comment (caiotomazelli) [inline dags/destinations/gads_oci.py:146]: Remove or improve this log

File: dags/destinations/gads_oci.py [hunk 4]
```diff
@@ -0,0 +1,353 @@
...
+        index = error.location.field_path_elements[0].index
+        message = f'Code: {error.error_code}, Error: {error.message}'
+        partial_failures[index] += message  # Can be multiple errors for the same conversion.
+
+    print(f"Partial failures: {partial_failures}")
+
+    return partial_failures
+
+  @staticmethod
+  def schema() -> Optional[ProtocolSchema]:
```
Comment (caiotomazelli) [inline dags/destinations/gads_oci.py:281]: The schema function should contain the fields that are expected to be present at the `config` object. I think the ones here refer to fields in the underlying customer data when it should really have the fields that you have under `_REQUIRED_GOOGLE_ADS_CREDENTIALS` and anything else you expect to be in the config object.

**Review by caiotomazelli — CHANGES_REQUESTED** [review]

File: airflow_requirements.txt (hunk 1 above)
Comment (gmcro) [reply airflow_requirements.txt]: Done, but the reason I didn't set it as I was getting some conflicts trying to generate the fixed package numbers. Let's test tomorrow to see if this works.

File: dags/destinations/gads_oci.py (hunk 4 above)
Comment (gmcro) [reply dags/destinations/gads_oci.py:276]: Done. Ohhh, I see. Confusion on my part.

File: airflow_requirements.txt (hunk 1 above)
Comment (gmcro) [reply airflow_requirements.txt]: Should be fixed now, but still seeing another integration test failure. Looks like it can't find integration_test.csvh.

**Review by jcryan-g — APPROVED** [review]
LGTM

**Review by caiotomazelli — APPROVED** [review]

**Pull request merged**